### DISCUSSION TITLE
Cloud Client C++ Libraries for Windows OS

### DISCUSSION THREAD

**Discussion by deepanshug** [googleapis/google-cloud-cpp#12713] [Q&A] [answered]
Hi, 

Are there any C++ libraries for Windows OS, I can see the ones available are for Linux only.
I am not even a beginner in Linux so asking if there is a way to use it in Windows?

**Comment by coryan** [comment, marked as answer]
> Hi,
> 
> Are there any C++ libraries for Windows OS,

We compile and test `google-cloud-cpp` on Windows (with MSVC), macOS, and Linux.

 > I can see the ones available are for Linux only.

What leads to you believe that?  I ask, because maybe some of our documentation or READMEs need fixing.

>  I am not even a beginner in Linux so asking if there is a way to use it in Windows?

Maybe this helps:  https://cloud.google.com/cpp/docs/setup#windows

**Comment by deepanshug** [comment]
First of all thanks for your quick response and apologies that I could not reply back earlier.
I followed the link you shared and was able to setup.
I simply installed google-cloud-cpp vcpkg and integrated it with visual studio. In visual studio, I can see all the headers available.
This solves my primary concern.

I tried to use this for my use case where I want to access google drive using service account.

I tried following options:
MakeJWTAssertion - Used the information from service-account.json file and could generate a JWT token but when checked in postman it says "Invalid JWT Signature".

However, in Python I used same information but code reads information directly from json file:
signer = crypt.RSASigner.from_service_account_file(sa_keyfile)

token generated in this case works well and returns me access token for further API calls.

Is there any similar way in C++ Client Library. I really could not find any documentation or example.

You can please guide me to relevant documentation for me to explore.

**Reply by dbolduc** [reply]
> I want to access google drive using service account.

FYI: We do not have C++ client libraries for Google Drive: https://github.com/googleapis/google-cloud-cpp/issues/3748

> generate a JWT token

We have helper functions to make `google::cloud::Credentials`...

https://cloud.google.com/cpp/docs/reference/common/latest/namespacegoogle_1_1cloud#group__guac_1ga0f2d23032efe73d17745a5f53112b918

The credentials can then be fed to `google::cloud::oauth2::MakeAccessTokenGenerator(...)`:

https://github.com/googleapis/google-cloud-cpp/blob/da7cecd860bb6b77c01f9da4de40e0610d1ee8c1/google/cloud/oauth2/access_token_generator.h#L74-L76

---

You are right that the documentation is lacking. We will work on that.

Here is the closest example, which passes the credentials to a Google Cloud Storage client: https://cloud.google.com/cpp/docs/reference/storage/latest/storage-auth-example

Putting it all together, the code would be something like:
```cc
std::string filename = ...

// Read service account file
auto is = std::ifstream(filename);
is.exceptions(std::ios::badbit);
auto json_string =
    std::string(std::istreambuf_iterator<char>(is.rdbuf()), {});

// Make credentials
auto credentials =
    google::cloud::MakeServiceAccountCredentials(json_string);

// Make access token generator from credentials
auto generator = google::cloud::oauth2::MakeAccessTokenGenerator(*credentials);

// Get access token
auto token = generator->GetToken();
if (!token) {
  std::cout << "Token creation failed with status: " << token.status() << std::endl;
} else {
  std::cout << "Created token: " << *token << std::endl;
}
```

**Reply by dbolduc** [reply]
Follow up to my previous comment:

The oauth2 library is not supported on its own in vcpkg. To enable it you will need to build a client library that depends on it. `storage` is one such client library.

e.g. `vcpkg install google-cloud-cpp[core,storage]`

https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/storage/quickstart/README.md#using-with-cmake

**Reply by coryan** [reply]
FWIW, the documentation and the example are now up:

https://cloud.google.com/cpp/docs/reference/oauth2/latest/namespacegoogle_1_1cloud_1_1oauth2
